app/controllers/restricted_proxy_controller.rb

```diff
@@ -1,12 +1,33 @@
 class RestrictedProxyController < ApplicationController
-  before_action :authenticate_user!
+  before_action :headers_auth
+
+  def headers_auth
+    JWT.decode(request.params['stacks_token'], Settings.geo_proxy_secret, true, { algorithm: 'HS256' })
+    set_cors_headers
+  rescue JWT::ExpiredSignature
+    render json: { error: 'Expired token' }, status: :unauthorized
+  rescue JWT::DecodeError
+    render json: { error: 'Invalid token' }, status: :unauthorized
+  end
+
+  def set_cors_headers
+    origin = request.origin
+    permitted_origins = [Settings.cors.allow_origin_url]
+    if permitted_origins.include?(origin)
+      response.headers['Access-Control-Allow-Origin'] = origin
+      response.headers['Access-Control-Allow-Credentials'] = true
+    else
+      response.headers['Access-Control-Allow-Origin'] = '*'
+    end
+  end
 
   def access
     proxied_headers.each do |k, v|
       headers[k] = v
     end
+
     self.status = proxied_response.status
-    self.response_body = proxied_response.body
+    send_data proxied_response.body, type: proxied_response.headers['Content-Type'], disposition: 'inline'
```

Comment:
`send_data` also takes a `status:` argument.  Can we combine line 29 into here using it?

Also, isn't `proxied_response` an instance of `HTTP::Response` https://www.rubydoc.info/github/httprb/http/HTTP/Response Does that have a `headers` method?  I don't see it in their docs.

Reply:
What headers are you talking about? The headers in set_cors_headers or the headers in    ``` proxied_headers.each do |k, v|
      headers[k] = v
    end```
    
 The first is for the controller. The second is what was previously in the code and I did not change it.

Reply:
I'm specifically speaking about `proxied_response.headers['Content-Type']`

Reply:
Also I double checked  proxied_response.headers and it returns #<HTTP::Headers:0x0000000154359738> so that is a valid field.

Reply:
Cool. Thanks for checking. Switching to an instance double 👇 would prevent doubts like this.

Reply:
I have done that.